docs/leases-and-lease-lifecycle.md

```diff
@@ -0,0 +1,154 @@
+# Leases and Lease Lifecycle
+
+Leases are created in, and leveraged by, KCL to manage a stream's shards.
+This document should help provide insights into the lease lifecycle.
+
+**Nota bene (N.B.):** while actual shard ids are formatted like `shardId-000000000042`, this document uses `shardId[_-]42` for concision.
+
+## Leases
+
+In KCL, a lease provides a temporal assignment between one Kinesis shard and an assigned worker.
+Leases are persistent for the duration of shard processing (detailed later).
+However, lease assignment is transient -- leases may be "stolen" by other workers in the same KCL application.
+
+## Lease Table
+
+To persist metadata about lease state (e.g., last read checkpoint, current assigned worker), KCL creates a lease table in [DynamoDB][dynamodb].
+Each KCL application will have its own distinct lease table that transcludes the application name.
+More information, including schema, is provided at [KCL LeaseTable][kcl-leasetable].
+
+## Lease Assignment
+
+The "life" of a lease closely mirrors the duration that a shard is being processed.
+
+![Activity diagram of KCL shard-to-lease assignments.
+shard-0 (lease-0) is unmodified.
+shard-1 (lease-1) is split into shard-4 (lease-4) and shard-5 (lease-5).
+shard-2 (lease-2) and shard-3 (lease-3) are merged into shard-6 (lease-6).
+](images/leases-and-mutations.png)
+
+Specifically, leases are unique to the shard and are not recycled for stream mutations (i.e., split, merge).
+Shards created by stream mutations will generate a new lease.
+
+It should be noted that the number of tracked leases may exceed the number of shards.
+Per the diagram (above), this can occur when there are stream mutations propagating through KCL.
+For example, a 10-shard stream that is split on every shard may temporarily have up-to 30 leases: 10 original + 20 children.
+
+N.B. Leases are uniquely identified by their `leaseKey` which looks vastly different than `lease_X`.
+For details on the `leaseKey` format, please see [KCL LeaseTable][kcl-leasetable].
+
+## Lease Lifecycle
+
+Leases follow a relatively simple, progressive state machine:
+`DISCOVERY -> CREATION -> PROCESSING -> SHARD_END -> DELETION`
```

Comment:
A quick side not that this is a very effective way of thinking, and I do think we should eventually codify it into the system.

Reply:
No objection from me! Resolving this conversation as that's outside the scope of this PR. :smile: